drivers: gpio: Add support for PCAL6408A I/O expander

anangl opened the pull request:
Add GPIO driver and devicetree binding for the PCAL6408A 8-bit I2C-bus I/O expander.
Include the driver in the build_all test.

Extracted from #31624.

anangl committed:
drivers: gpio: Add support for PCAL6408A I/O expander

Add GPIO driver and devicetree binding for the PCAL6408A 8-bit I2C-bus
I/O expander.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
tests: drivers: build_all: Add PCAL6408A GPIO driver

Include the PCAL6408A I/O expander driver in the build_all test.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

mbolivar-nordic reviewed (changes requested):
Looks good! Only the device_is_ready() comment is blocking.

mnkp commented on dts/bindings/gpio/nxp,pcal6408a.yaml:
```diff
@@ -0,0 +1,51 @@
+      description: |
+        GPIO connected to the controller RESET pin. This pin is active-low.
+
+    init-out-low:
+      type: int
+      required: false
+      default: 0
+      description: |
+        Bit mask identifying pins that should be initialized as outputs
+        driven low.
+
+    init-out-high:
+      type: int
+      required: false
+      default: 0
+      description: |
+        Bit mask identifying pins that should be initialized as outputs
+        driven high.
```
We want to base our DT on the Linux implementation. This functionality is implemented there using GPIO hogging mechanism as described in [GPIO device tree bindings](https://github.com/torvalds/linux/blob/7a7fd0de4a9804299793e564a555a49c1fc924cb/Documentation/devicetree/bindings/gpio/gpio.txt#L185) document. We haven't considered yet how this could be implemented in Zephyr. The above approach is however non-standard.

anangl replied:
I copied these two properties from the ["semtech,sx1509b" binding](https://github.com/zephyrproject-rtos/zephyr/blob/48321ef2525c14c4d9b6040afe8ae72ff07a8d1e/dts/bindings/gpio/semtech%2Csx1509b-gpio.yaml#L28-L42), but I actually don't use this functionality, so I'm happy to drop it. It can be added later if needed.

mnkp replied:
This functionality is generally useful but I propose to drop it for now and implement later using GPIO hogging mechanism.

mnkp reviewed (commented):
LGTM. Could you test the driver with `tests/drivers/gpio/gpio_api_1pin/` and `tests/drivers/gpio/gpio_basic_api/` testcase?

anangl committed:
tests: drivers: build_all: Add fake instance of GPIO device

Such instance is needed for building drivers that use DEVICE_DT_GET()
to get the pointer to the GPIO controller device structure.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented on drivers/gpio/gpio_pcal6408a.c:
```diff
@@ -0,0 +1,648 @@
+	 * this driver.
+	 */
+	if (drv_cfg->reset_gpio_dev) {
+		rc = gpio_pin_configure(drv_cfg->reset_gpio_dev,
```
Thanks for catching this. Added.

anangl commented on drivers/gpio/gpio_pcal6408a.c:
```diff
@@ -0,0 +1,648 @@
+			.port_pin_mask =				   \
+				GPIO_PORT_PIN_MASK_FROM_DT_INST(idx),	   \
+		},							   \
+		.i2c = DEVICE_DT_GET(DT_BUS(DT_DRV_INST(idx))),		   \
```
Applied.

mnkp commented:
Is the `tests/drivers/gpio/gpio_api_1pin/` passing? It does not require any setup, it's enough to flash and run it.

anangl commented:
> Is the `tests/drivers/gpio/gpio_api_1pin/` passing? It does not require any setup, it's enough to flash and run it.

Yes, it passes now (I had to add the indication that simultaneous input/output mode is not supported, as this expander can generate interrupts only for input pins ):
```
*** Booting Zephyr OS build zephyr-v2.5.0-590-g031e7e137041  ***
Running test suite gpio_api_1pin_test
===================================================================
START - test_gpio_pin_configure_push_pull
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_pin_configure_push_pull
===================================================================
START - test_gpio_pin_configure_single_ended
Running test on port=GPIO_P0, pin=4
When configured as input test pin value is low
Open Source configuration or Pull Down pin bias is not supported
 SKIP - test_gpio_pin_configure_single_ended
===================================================================
START - test_gpio_pin_set_get_raw
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_pin_set_get_raw
===================================================================
START - test_gpio_pin_set_get
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_pin_set_get
=====================================================
START - test_gpio_pin_set_get_active_high
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_pin_set_get_active_high
===================================================================
START - test_gpio_pin_set_get_active_low
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_pin_set_get_active_low
===================================================================
START - test_gpio_pin_toggle
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_pin_toggle
===================================================================
START - test_gpio_port_set_masked_get_raw
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_port_set_masked_get_raw
===================================================================
START - test_gpio_port_set_masked_get
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_port_set_masked_get
===================================================================
START - test_gpio_port_set_masked_get_active_high
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_port_set_masked_get_active_high
===================================================================
START - test_gpio_port_set_masked_get_active_low
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_port_set_masked_get_active_low
===================================================================
START - test_gpio_port_set_bits_clear_bits_raw
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_port_set_bits_clear_bits_raw
===================================================================
START - test_gpio_port_set_bits_clear_bits
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_port_set_bits_clear_bits
===================================================================
START - test_gpio_port_set_clr_bits_raw
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_port_set_clr_bits_raw
===================================================================
START - test_gpio_port_set_clr_bits
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_port_set_clr_bits
===================================================================
START - test_gpio_port_toggle
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_port_toggle
===================================================================
START - test_gpio_int_edge_rising
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_int_edge_rising
===================================================================
START - test_gpio_int_edge_falling
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_int_edge_falling
===================================================================
START - test_gpio_int_edge_both
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_int_edge_both
===================================================================
START - test_gpio_int_edge_to_active
Step 1: Configure pin as active high
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_int_edge_to_active
===================================================================
START - test_gpio_int_edge_to_inactive
Step 1: Configure pin as active high
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_int_edge_to_inactive
===================================================================
START - test_gpio_int_level_high
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_int_level_high
===================================================================
START - test_gpio_int_level_low
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_int_level_low
===================================================================
START - test_gpio_int_level_active
Step 1: Configure pin as active high
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_int_level_active
===================================================================
START - test_gpio_int_level_inactive
Step 1: Configure pin as active high
Running test on port=GPIO_P0, pin=4
Simultaneous pin in/out mode is not supported.
 SKIP - test_gpio_int_level_inactive
===================================================================
START - test_gpio_pin_toggle_visual
Running test on port=GPIO_P0, pin=4
LED ON
LED OFF
LED ON
LED OFF
 PASS - test_gpio_pin_toggle_visual
===================================================================
Test suite gpio_api_1pin_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

I also tried to get `tests/drivers/gpio/gpio_basic_api/` to work, but unfortunately it is not possible on the board that I have (nRF9160DK 0.15.1) without actually destroying it (all pins of this expander are connected to LEDs and buttons on that board, with pull-downs and pull-ups, respectively). I tested generation of interrupts on `samples/basic/button` and I actually found one more [thing that had to be corrected](https://github.com/zephyrproject-rtos/zephyr/compare/db0c682ec216eecdb1dbebd12e80ae7e57e57033..3ee98c457774c66c0929d73cf658b969d059e8bf#diff-53a2b120dbd79905bffc2a4a01d5081228f78c30636f1c453192b088e68b0b7bR233).

pabigot reviewed (approved):
Can't test, but looks reasonable.
